File: src/Y2015/D02.hs
module Y2015.D02 where

import System.Environment (getArgs)
import Data.List
import Data.List.Split (splitOn)
import Control.Monad (liftM2, ap)

area :: [Int] -> Int
area [l,w,h] = 2*l*w + 2*w*h + 2*h*l
area _ = error "invalid dimensions"

smallest :: [Int] -> [Int]
smallest = take 2 . sort

paper :: [Int] -> Int
paper = liftM2 (+) area (product . smallest)

ribbon :: [Int] -> Int
ribbon = ap ((+) . sum . map (2 *) . smallest) product
-- ^ The bow is the product and the rest is the perimter

parseLine :: String -> [Int]
parseLine = map (read::String->Int) . splitOn "x"

part1 :: [String] -> Int
part1 = sum . map (paper . parseLine)

part2 :: [String] -> Int
part2 = sum . map (ribbon . parseLine)

main :: IO ()
main = do
   args <- getArgs
   content <- readFile (head args)
   let l = lines content
   putStrLn $ "Part 1: " ++ show (part1 l)
   putStrLn $ "Part 1: " ++ show (part2 l)

File: python/2015/d02.py
import sys

def area(l, w, h):
    return 2*l*w + 2*w*h + 2*h*l

def smallest(l, w, h):
    return sorted((l, w, h))[:2]

def paper(l, w, h):
    x = smallest(l, w, h)
    return area(l, w, h) + x[0] * x[1]

def ribbon(l, w, h):
    x = smallest(l, w, h)
    perim = x[0]*2 + x[1]*2
    bow = l * w * h
    return perim + bow


with open(sys.argv[1], 'r') as f:
    lines = f.read().splitlines()
    total = 0
    for line in lines:
        l, w, h = [int(x) for x in line.split('x')]
        total += paper(l, w, h)
    print 'Part 1:', total
    total = 0
    for line in lines:
        l, w, h = [int(x) for x in line.split('x')]
        total += ribbon(l, w, h)
    print 'Part 2:', total
